File: mac/Sources/FineClipboard/Services/SyncClient.swift
import Foundation

/// Thin async HTTP client for the finepaste-server sync API.
final class SyncClient {
    var baseURL: String = ""
    var token: String?

    struct SyncError: LocalizedError {
        let message: String
        var errorDescription: String? { message }
    }

    struct AuthResult: Codable {
        let token: String
        let vip: Bool
        let vip_until: Int64?
    }
    struct PushItem: Codable {
        let uuid: String
        let kind: Int
        let updated_at: Int64
        let deleted: Bool
        let cipher: String
    }
    struct PushResult: Codable { let cursor: Int64 }
    struct Change: Codable {
        let uuid: String
        let kind: Int
        let seq: Int64
        let updated_at: Int64
        let deleted: Bool
        let blob_size: Int
        let cipher: String?
    }
    struct ChangesResult: Codable {
        let cursor: Int64
        let has_more: Bool
        let changes: [Change]
    }

    func register(email: String, password: String) async throws -> AuthResult {
        try await postJSON("/api/register", body: ["email": email, "password": password])
    }
    func login(email: String, password: String) async throws -> AuthResult {
        try await postJSON("/api/login", body: ["email": email, "password": password])
    }
    func redeem(key: String) async throws -> AuthResult {
        try await postJSON("/api/redeem", body: ["key": key], authed: true)
    }

    func push(retentionDays: Int, items: [PushItem]) async throws -> PushResult {
        var req = try request("POST", "/api/sync/push", authed: true)
        let body = PushBody(retention_days: retentionDays, items: items)
        req.httpBody = try JSONEncoder().encode(body)
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        return try await send(req)
    }

    func changes(since: Int64, limit: Int = 200) async throws -> ChangesResult {
        let req = try request("GET", "/api/sync/changes?since=\(since)&limit=\(limit)", authed: true)
        return try await send(req)
    }

    func blob(uuid: String) async throws -> Data? {
        let req = try request("GET", "/api/sync/blob/\(uuid)", authed: true)
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse else { return nil }
        if http.statusCode == 404 { return nil }
        guard (200..<300).contains(http.statusCode) else { throw SyncError(message: "HTTP \(http.statusCode)") }
        return data
    }

    // MARK: - helpers

    private struct PushBody: Codable {
        let retention_days: Int
        let items: [PushItem]
    }

    private func request(_ method: String, _ path: String, authed: Bool = false) throws -> URLRequest {
        guard let url = URL(string: baseURL + path) else { throw SyncError(message: "无效的服务器地址") }
        var req = URLRequest(url: url)
        req.httpMethod = method
        if authed, let token { req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        return req
    }

    private func postJSON<T: Decodable>(_ path: String, body: [String: String], authed: Bool = false) async throws -> T {
        var req = try request("POST", path, authed: authed)
        req.httpBody = try JSONEncoder().encode(body)
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        return try await send(req)
    }

    private func send<T: Decodable>(_ req: URLRequest) async throws -> T {
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse else { throw SyncError(message: "无响应") }
        guard (200..<300).contains(http.statusCode) else {
            throw SyncError(message: errorMessage(data) ?? "HTTP \(http.statusCode)")
        }
        return try JSONDecoder().decode(T.self, from: data)
    }

    private func errorMessage(_ data: Data) -> String? {
        struct E: Codable { let error: String? }
        return (try? JSONDecoder().decode(E.self, from: data))?.error
    }
}

File: src/FineClipboard/Services/SyncClient.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Http;
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

namespace FineClipboard.Services;

/// <summary>Thin HTTP client for the finepaste-server sync API.</summary>
public sealed class SyncClient
{
    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(30) };

    public string BaseUrl { get; set; } = string.Empty;
    public string? Token { get; set; }

    public sealed class AuthResult
    {
        [JsonPropertyName("token")] public string Token { get; set; } = string.Empty;
        [JsonPropertyName("vip")] public bool Vip { get; set; }
        [JsonPropertyName("vip_until")] public long? VipUntil { get; set; }
    }

    public sealed class PushItem
    {
        [JsonPropertyName("uuid")] public string Uuid { get; set; } = string.Empty;
        [JsonPropertyName("kind")] public int Kind { get; set; }
        [JsonPropertyName("updated_at")] public long UpdatedAt { get; set; }
        [JsonPropertyName("deleted")] public bool Deleted { get; set; }
        [JsonPropertyName("cipher")] public string Cipher { get; set; } = string.Empty;
    }

    public sealed class PushResult
    {
        [JsonPropertyName("cursor")] public long Cursor { get; set; }
    }

    public sealed class Change
    {
        [JsonPropertyName("uuid")] public string Uuid { get; set; } = string.Empty;
        [JsonPropertyName("kind")] public int Kind { get; set; }
        [JsonPropertyName("seq")] public long Seq { get; set; }
        [JsonPropertyName("updated_at")] public long UpdatedAt { get; set; }
        [JsonPropertyName("deleted")] public bool Deleted { get; set; }
        [JsonPropertyName("blob_size")] public int BlobSize { get; set; }
        [JsonPropertyName("cipher")] public string? Cipher { get; set; }
    }

    public sealed class ChangesResult
    {
        [JsonPropertyName("cursor")] public long Cursor { get; set; }
        [JsonPropertyName("has_more")] public bool HasMore { get; set; }
        [JsonPropertyName("changes")] public List<Change> Changes { get; set; } = new();
    }

    public Task<AuthResult?> RegisterAsync(string email, string password) =>
        PostAuthAsync("/api/register", email, password);

    public Task<AuthResult?> LoginAsync(string email, string password) =>
        PostAuthAsync("/api/login", email, password);

    private async Task<AuthResult?> PostAuthAsync(string path, string email, string password)
    {
        using HttpResponseMessage resp = await Http.PostAsJsonAsync(
            BaseUrl + path, new { email, password }).ConfigureAwait(false);
        if (!resp.IsSuccessStatusCode)
        {
            throw new SyncException(await ErrorMessageAsync(resp).ConfigureAwait(false));
        }
        return await resp.Content.ReadFromJsonAsync<AuthResult>().ConfigureAwait(false);
    }

    public async Task<AuthResult?> RedeemAsync(string key)
    {
        using HttpRequestMessage req = Authed(HttpMethod.Post, "/api/redeem");
        req.Content = JsonContent.Create(new { key });
        using HttpResponseMessage resp = await Http.SendAsync(req).ConfigureAwait(false);
        if (!resp.IsSuccessStatusCode)
        {
            throw new SyncException(await ErrorMessageAsync(resp).ConfigureAwait(false));
        }
        return await resp.Content.ReadFromJsonAsync<AuthResult>().ConfigureAwait(false);
    }

    public async Task<PushResult> PushAsync(int retentionDays, IEnumerable<PushItem> items)
    {
        using HttpRequestMessage req = Authed(HttpMethod.Post, "/api/sync/push");
        req.Content = JsonContent.Create(new { retention_days = retentionDays, items });
        using HttpResponseMessage resp = await Http.SendAsync(req).ConfigureAwait(false);
        if (!resp.IsSuccessStatusCode)
        {
            throw new SyncException(await ErrorMessageAsync(resp).ConfigureAwait(false));
        }
        return (await resp.Content.ReadFromJsonAsync<PushResult>().ConfigureAwait(false))!;
    }

    public async Task<ChangesResult> ChangesAsync(long since, int limit = 200)
    {
        using HttpRequestMessage req = Authed(HttpMethod.Get, $"/api/sync/changes?since={since}&limit={limit}");
        using HttpResponseMessage resp = await Http.SendAsync(req).ConfigureAwait(false);
        if (!resp.IsSuccessStatusCode)
        {
            throw new SyncException(await ErrorMessageAsync(resp).ConfigureAwait(false));
        }
        return (await resp.Content.ReadFromJsonAsync<ChangesResult>().ConfigureAwait(false))!;
    }

    public async Task<byte[]?> BlobAsync(string uuid)
    {
        using HttpRequestMessage req = Authed(HttpMethod.Get, $"/api/sync/blob/{uuid}");
        using HttpResponseMessage resp = await Http.SendAsync(req).ConfigureAwait(false);
        if (resp.StatusCode == HttpStatusCode.NotFound)
        {
            return null;
        }
        if (!resp.IsSuccessStatusCode)
        {
            throw new SyncException(await ErrorMessageAsync(resp).ConfigureAwait(false));
        }
        return await resp.Content.ReadAsByteArrayAsync().ConfigureAwait(false);
    }

    private HttpRequestMessage Authed(HttpMethod method, string path)
    {
        var req = new HttpRequestMessage(method, BaseUrl + path);
        if (!string.IsNullOrEmpty(Token))
        {
            req.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", Token);
        }
        return req;
    }

    private static async Task<string> ErrorMessageAsync(HttpResponseMessage resp)
    {
        try
        {
            string body = await resp.Content.ReadAsStringAsync().ConfigureAwait(false);
            using JsonDocument doc = JsonDocument.Parse(body);
            if (doc.RootElement.TryGetProperty("error", out JsonElement e))
            {
                return e.GetString() ?? resp.StatusCode.ToString();
            }
        }
        catch
        {
            // fall through
        }
        return $"HTTP {(int)resp.StatusCode}";
    }
}

/// <summary>A sync operation failed; Message carries a user-facing reason.</summary>
public sealed class SyncException : Exception
{
    public SyncException(string message) : base(message) { }
}
